apps/fhir/server/authentication.py

```diff
@@ -8,31 +10,153 @@
+def log_fhir_id_not_matched(mbi_hash, hicn_hash,
+                            hash_lookup_type, hash_lookup_mesg):
+    '''
+        Logging for "FhirIDNotFound" type
+        used in match_backend_patient_identifier()
+    '''
+    logger.info(json.dumps({
+        "type": "FhirIDNotFound",
+        "mbi_hash": mbi_hash,
+        "hicn_hash": hicn_hash,
+        "hash_lookup_type": hash_lookup_type,
+        "hash_lookup_mesg": hash_lookup_mesg,
+    }))
+
+
+def log_fhir_id_matched(fhir_id, mbi_hash, hicn_hash,
+                        hash_lookup_type, hash_lookup_mesg):
+    '''
+        Logging for "FhirIDFound" type
+        used in match_backend_patient_identifier()
+    '''
+    logger.info(json.dumps({
+        "type": "FhirIDFound",
+        "fhir_id": fhir_id,
+        "mbi_hash": mbi_hash,
+        "hicn_hash": hicn_hash,
+        "hash_lookup_type": hash_lookup_type,
+        "hash_lookup_mesg": hash_lookup_mesg,
+    }))
+
+
+def match_backend_patient_identifier(mbi_hash, hicn_hash):
+    '''
+    Matches a patient identifier via the backend FHIR server
+    using an MBI or HICN hash.
+
+    Summary:
+        1. mbi_hash is used for the initial lookup.
+        2. If there is a mbi lookup issue, the hicn_hash is used next.
+        3. If there is a hicn lookup issue, exceptions are raised.
+        4. A NotFound exception is raised if no match was found.
+
+    Returns:
+        fhir_id = Matched patient identifier.
+        backend_data = Passed thru to caller. Utilized in TestAuthentication.
+        hash_lookup_type = The type used for the successful lookup (M or H).
+
+    Raises:
+        UpstreamServerException: If hicn_hash search found duplicates.
+        NotFound: If both searches did not match a fhir_id.
+    '''
     auth_state = FhirServerAuth(None)
     certs = (auth_state['cert_file'], auth_state['key_file'])
-    # URL for patient ID.
+
+    # 1. mbi_hash is used for the initial lookup.
+    hash_lookup_type = "M"
+    hash_lookup_mesg = None
+
+    if mbi_hash is not None:
+        # URL for patient ID search by mbi_hash.
+        url = get_resourcerouter().fhir_url + \
+            "Patient/?identifier=" + \
+            settings.FHIR_SEARCH_PARAM_IDENTIFIER_MBIHASH + "%7C" + \
+            mbi_hash + \
+            "&_format=" + \
+            settings.FHIR_PARAM_FORMAT
+
+        response = requests.get(url, cert=certs, verify=False)
+        response.raise_for_status()
+        backend_data = response.json()
+
+        # Check resource bundle total > 1
+        if backend_data.get('total', 0) > 1:
+            hash_lookup_mesg = "Duplicate beneficiaries found in Patient resource bundle total"
+
+        # Check number of resource entries > 1
+        if (
+            'entry' in backend_data
+            and len(backend_data['entry']) > 1
+            and hash_lookup_mesg is not None
+        ):
+            hash_lookup_mesg = "Duplicate beneficiaries found in Patient resource bundle entry"
```

Comment:
Can you explain the difference in duplicate scenarios here?  Just curious since they return the same lookup message, if they could be combined into the same if statement.

Reply:
I see the HICN hash lookup duplicate checks are structured the same way so no issue, just for my own knowledge.

Reply:
@njdister One of the checks looks at the total returned in the response. The other check verifies that by counting the actual entries.

Reply:
Ah, the old trust but verify, got it!